Add a SleepAction / Handler (for pod lifecycle / preStop)

rtreffer opened the pull request:
A sleep action is mainly useful during shutdown to delay the delivery of the TERM/KILL signal.

**What this PR does / why we need it**:

Delaying the SIGTERM by a few seconds helps to drain pods. Other methods (tcp, http, exec) are tied to the pod or software and are thus not applicable in all situations.

I've added a longer description (including alternatives) to #64510 - this PR is one possible solution.

Usage:
```
        lifecycle:
          preStop:
            sleep:
              seconds: 15
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #64510

**Special notes for your reviewer**:

I've tested this with a local testcluster and it worked as intended.

The [implementation](https://github.com/kubernetes/kubernetes/compare/master...soundcloud:sleep-handler?expand=1#diff-9a9f71a82ddbbcd0627de3aea90c6649R73) is a three-liner, everything else is an extension to the configuration format and autogenerated code (deepcopy/protobuf/...)

I am happy to throw this away and take a different route if needed. Or to fix whatever issues come up. Or to discuss other alternative.

**Release note**:
```release-note
A sleep handler was added as an option for lifecycle hooks. The handler can be used to delay the SIGTERM send to the container.
```

/sig node

dims commented:
/unassign
/uncc

dims commented:
/ok-to-test

rtreffer commented:
/test pull-kubernetes-e2e-gce

rtreffer commented:
/assign @thockin

As noted in the initial issue #64510 there are other options as well. This is a simple sleep lifecycle hook to delay the SIGTERM by a few seconds (which helps a lot with error rates due to pod shutdowns).

thockin commented:
Let's consider the alternatives.

You could wrap your binary in a script that catches SIGTERM and either ignores it or converts it to another signal.  Not great if you want to run someone else's image without building a derivative.

We could add a `terminationSignal` field, default = SIGTERM, and send that instead.  This is what docker does with STOPSIGNAL in Dockerfiles.

We could add this here.  Since you only get one Lifecycle hook, this is less than ideal in my mind.

Other options?

rtreffer commented:
@thockin totally agree, this is more a question of "what's a good way to do this" than an actual implementation. I have listed some thoughts/alternatives in #64510

I would add the options I listed in #64510:
1. Add a new pod option - this would overcome the limit of one hook.
1. Add an option to the kubelet - as we know our terminating propagation time it might be an option to set that cluster-wide through kubelet flags

**Our motivation for a sleep action** (the reaul user facing issue):
Whenever we did a rolling cluster upgrade / reprovisioning user-facing errors were elevated and regularly spiking. The main reason were some pods that instantly stopped while other pods where still sending traffic there. Our simple workaround was the addition of a `sleep 15` (15 = 3 * DNS TTL) exec shutdown hook by default.
The result were way less errors during cluster upgrades and more precise metrics as there is enough time for a last post-traffic scrape. But we ran into a new issue where pods would sometimes enter a broken state (`/proc` unmounted) on termination. This state could not be resolved by kubernetes and requires operator intervention. (This is likely another bug but it happens only on our most-complex pod setup an we haven't isolated this yet)

So what we would like to solve:
1. Makes it easy to add a graceful termination period
2. Ideally decoupled from the pod / container
@thockin making the termination signal configurable (or even turning it off) should work. Although it would be hard to find a good default (except for "off").

I am happy to throw this PR away and implement whatever looks like a good solution.

schweikert commented:
Regarding the "only one hook" issue. Wouldn't it be possible to allow specifying both a "exec" and a "sleep" parameter to preStop, and respect both settings by first sleeping and then executing the command?

        lifecycle:
          preStop:
            sleep:
             seconds: 15
            exec:
              command: ["/stop-service.sh"]

rtreffer commented:
@schweikert this is part of the validation check in https://github.com/kubernetes/kubernetes/pull/64888/files#diff-81c0c668d90c62bb5b7a2334c09c7e50R2482 and could be lifted.

- Add a sleep hook that is always allowed (lift validation limit) and always executed first (by convention)

yujuhong commented:
> Let's consider the alternatives.
> 
> You could wrap your binary in a script that catches SIGTERM and either ignores it or converts it to another signal. Not great if you want to run someone else's image without building a derivative.

If you're okay with building your own image, you can simply specify the STOPSIGNAL in your dockerfile. Kubernetes (at least w/ docker or containerd) respects the stop signal in the image, so this'd work today (see https://github.com/kubernetes/kubernetes/issues/30039#issuecomment-370082283).

> 
> We could add a terminationSignal field, default = SIGTERM, and send that instead. This is what docker does with STOPSIGNAL in Dockerfiles.

This is an option. We'll need to add the field to CRI too to support the feature.

> 
> We could add this here. Since you only get one Lifecycle hook, this is less than ideal in my mind.
> 

I don't see this as a huge problem. The primary purpose of prestop is to offer application-specific shutdown behavior. The sleep action introduced in this PR is a "simple solution" for pods which doesn't need any other custom actions. If a custom prestop hook is required for the pod, it's likely that you don't need the simply sleep action, or you could include the sleep time in your hook as well.

> Other options?

thockin commented on pkg/kubelet/lifecycle/handlers.go:
```diff
@@ -69,6 +70,9 @@ func (hr *HandlerRunner) Run(containerID kubecontainer.ContainerID, pod *v1.Pod,
 		}
 		return msg, err
+	case handler.Sleep != nil:
+		time.Sleep(time.Duration(int64(handler.Sleep.Seconds) * int64(time.Second)))
```
This will cause the calling code to block, potentially for a very long time.  That changes the semantics of how and where it can be used quite a lot.

If we're literally going to sleep here, we need to do a through examination of the entire callstack (all the potential callstacks, actually) to be sure it is safe to block.  I don't know Kubelet's internals well enough any more to say for sure, but a quick glance at callers seems to suggest that this would block `Kubelet.killPod` which DOES seem to be run from a dedicated goroutine.  So this may be OK but:

a) @yujuhong needs to confirm
b) This should be commented about here

rtreffer replied:
Hm, it does not make sense to sleep longer than the killPod signal. So we could potentially limit sleep to the deadline of the pod.

At least the `exec` hook has the same issue of potentially unbound runtime (well, up to killPod).

thockin replied:
ping @yujuhong

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1967,6 +1974,9 @@ type Handler struct {
 	TCPSocket *TCPSocketAction
+	// Sleep specifies a wait time.
+	// +optional
+	Sleep *SleepAction
```
Needs better comments about why someone one use this and what the effect is WRT signal delivery.

thockin replied:
Handler is used in other places.  This means that `sleep` is now an option for livenessProbes, for readinessProbes, for postStart hooks, etc.  I did a very cursory analysis below of how this is PROBABLY safe for preStop, but I doubt we can (or want to) make the same guarantees about every user of Handler.

Aside from that, can we make even a hypothetical case for why sleep would make sense in any of those?

Maybe better to make a new struct, embed `Handler`, and add `sleep`; or just clone all of `Handler` into `preStopHandler`. And then fix all the related code.

rtreffer replied:
The only hypothetical use I could see for sleep in the other cases would be to build test cases / build reproduce cases.

thockin replied:
Agree - we should fix this to only affect preStop

rtreffer replied:
I tried copying the whole struct but then noticed that I would need to duplicate some code in pkg/kubelet/lifecycle/handlers.go. I'll go with embedding path to avoid that.

rtreffer replied:
Hm, I just tried that, but can this work with existing clusters? Wouldn't the new definition break existing clusters?

Looks like I should go back to copying the struct.

thockin replied:
Not sure what you mean breaking clusters, wrt Handler.

https://play.golang.org/p/u-JoOYkQC9o

This *could*, however break clusters with version skew.  Validation today demands that exactly one Handler be defined, and a kubelet could be back-rev vs an apiserver.

This is a general problem and has come up repeatedly.  Maybe we should do a release where we remove that validation pattern wherever we can find it, and then these changes would be less awful.

Otherwise we force a 2-release window on ever such change.

@liggitt @bgrant0607

thockin commented on pkg/apis/core/types.go:
```diff
@@ -1807,6 +1807,13 @@ type ExecAction struct {
 
+// SleepAction describes a sleep action.
+type SleepAction struct {
+	// The sleep duration in seconds.
```
needs comments about bounds

thockin replied:
Why is it optional?  Can we imagine a Sleep struct without it?

rtreffer replied:
Most of the other fields are optional, I guessed because it goes into the protobuf and protobuf has this big warning about `Required Is Forever`.
But then again I can't imagine it to be missing, I'll drop it.

thockin reviewed (commented):
I guess I could live with this.

But there's no such thing as a 3-liner :)

rtreffer commented:
@thockin rebased on master, limited to 60s sleep which should be safe and enough for DNS based draining (would be great to get a verification from @yujuhong ).

Test failures seem to be unrelated (timeout and missing logs?).

rtreffer commented:
/retest

rtreffer commented:
/retest

rtreffer commented:
/test pull-kubernetes-integration

rtreffer commented:
/retest

rtreffer commented:
@yujuhong / @thockin can I have another check?

Main changes:
- limit of 60s for the wait time - this should solve concerns around unbound wait time and longer wait times should be done in a cooperative way with the container userspace
- more doc blocks
- rebase

rtreffer commented:
/retest

rtreffer commented:
/retest

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2013,6 +2019,12 @@ type Handler struct {
 	TCPSocket *TCPSocketAction
+	// Sleep specifies a wait time that will be executed in the kubelet only.
+	// Using sleep is useful during pod shutdown to delay the SIGTERM and thus
+	// given DNS and API clients time to drain the pod before the application
```
s/given/give

thockin commented on pkg/apis/core/validation/validation.go:
```diff
@@ -2485,6 +2485,16 @@ func validateTCPSocketAction(tcp *core.TCPSocketAction, fldPath *field.Path) fie
+func validateSleepAction(sleep *core.SleepAction, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	if sleep.Seconds <= 1 || sleep.Seconds >= 60 {
```
define a const for 60 please

rtreffer-sc committed:
Add a SleepAction for the pod lifecycle.

A sleep action is mainly useful during the shutdown hook to delay the
delivery of the TERM signal.

rtreffer-sc committed:
Copy Handler fields into PreStopHandler

rtreffer commented:
/retest

rtreffer commented:
There is now a dedicated PreStopHandler struct. It has the same fields as the Handler struct plus the sleep action. Protobuf field definitions are identical, so it should deserialize pre-PR versions without issues (please correct me if I'm wrong here).

There is some duplication in the validation as the "exactly one handler action" rule is a bit hard to verify without looking at all fields of the struct.

thockin commented on pkg/apis/core/types.go:
```diff
@@ -2015,6 +2021,27 @@ type Handler struct {
 
+// PreStopHandler defines the action that can be taken before a Pod is shut down.
+type PreStopHandler struct {
+	// One and only one of the following should be specified.
```
Why not embed Handler here?  e.g.

```
type PreStopHandler struct {
    Handler
  // comments...
    Sleep *SleepAction
}

bgrant0607 commented:
This really needs a KEP to discuss the design and implications.

https://github.com/kubernetes/enhancements/blob/master/keps/0001-kubernetes-enhancement-proposal-process.md

liggitt commented:
/hold
for KEP
